File: Problems/0115. Distinct Subsequences/solution2.py
# DP Top-Down
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        dp = {}
        
        def dfs(i, j):
            if (i, j) in dp:
                return dp[(i, j)]
            if j == len(t):
                return 1
            elif i == len(s):
                return 0

            dp[(i, j)] = dp.get((i, j), 0) + dfs(i + 1, j) + (dfs(i + 1, j + 1) if s[i] == t[j] else 0)

            return dp[(i, j)]

        return dfs(0, 0)

File: Problems/0230. Kth Smallest Element In a Bst/solution2.kt
# DFS
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        result = -1
        
        def inorder(node, count):
            nonlocal result
            if count == k: return count
            elif not node: return count

            count = inorder(node.left, count) + 1
            if count == k:
                result = node.val
                return count

            count = inorder(node.right, count)
            return count

        inorder(root, 0)
        return result
